File: arena-backend/src/modules/session/session.service.ts
import { Inject, Injectable, OnModuleInit, forwardRef } from '@nestjs/common';
import Redis from 'ioredis';
import { REDIS_CLIENT } from 'src/redis/redis.constants';
import { Signal } from 'src/signal/signal';
import { SignalChannel } from 'src/signal/signal.channels';
import { UserService } from '../user/user.service';
import { CachedUser } from './session.types';

const SESSION_PREFIX = 'session:';
const SESSION_TTL = 300; // 5분

@Injectable()
export class SessionService implements OnModuleInit {
  constructor(
    @Inject(REDIS_CLIENT)
    private readonly redis: Redis,
    private readonly signal: Signal,
    @Inject(forwardRef(() => UserService))
    private readonly userService: UserService,
  ) {}

  onModuleInit() {
    this.signal.subscribe(SignalChannel.USER_UPDATED, async ({ uid }) => {
      await this.invalidate(uid);
    });
  }

  async getOrFetch(uid: string): Promise<CachedUser | null> {
    // 1. 캐시에서 조회
    let cachedUser = await this.get(uid);

    // 2. 캐시 미스 → DB 조회 후 캐싱
    if (!cachedUser) {
      const userEntity = await this.userService.findByUid(uid);
      if (userEntity) {
        cachedUser = {
          userId: userEntity.userId,
          uid: userEntity.uid,
          utag: userEntity.utag,
          nick: userEntity.nick,
          email: userEntity.email,
        };
        await this.set(cachedUser);
      }
    }

    return cachedUser;
  }

  async get(uid: string): Promise<CachedUser | null> {
    const key = this.getKey(uid);
    const data = await this.redis.get(key);
    if (!data) {
      return null;
    }
    return JSON.parse(data) as CachedUser;
  }

  async set(user: CachedUser): Promise<void> {
    const key = this.getKey(user.uid);
    await this.redis.set(key, JSON.stringify(user), 'EX', SESSION_TTL);
  }

  async invalidate(uid: string): Promise<void> {
    const key = this.getKey(uid);
    await this.redis.del(key);
  }

  private getKey(uid: string): string {
    return `${SESSION_PREFIX}${uid}`;
  }
}

File: legacy/backend/src/main/java/app/sandori/arena/api/domain/session/SessionService.java
package app.sandori.arena.api.domain.session;

import app.sandori.arena.api.domain.user.UserEntity;
import app.sandori.arena.api.domain.user.UserService;
import app.sandori.arena.api.global.signal.Signal;
import app.sandori.arena.api.global.signal.SignalChannel;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import java.util.concurrent.TimeUnit;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class SessionService {

    private static final String SESSION_PREFIX = "session:";
    private static final long SESSION_TTL = 300;

    private final StringRedisTemplate redisTemplate;
    private final Signal signal;
    private final UserService userService;
    private final ObjectMapper objectMapper = new ObjectMapper();

    @PostConstruct
    void init() {
        signal.subscribe(SignalChannel.USER_UPDATED, data -> {
            try {
                JsonNode node = objectMapper.readTree(data);
                String uid = node.get("uid").asText();
                invalidate(uid);
            } catch (Exception e) {
                log.error("Failed to handle USER_UPDATED signal", e);
            }
        });
    }

    public CachedUser getOrFetch(String uid) {
        CachedUser cached = get(uid);
        if (cached != null) {
            return cached;
        }

        UserEntity userEntity = userService.findByUid(uid);
        if (userEntity != null) {
            CachedUser cachedUser = new CachedUser(
                    userEntity.getUserId(),
                    userEntity.getUid(),
                    userEntity.getUtag(),
                    userEntity.getNick(),
                    userEntity.getEmail()
            );
            set(cachedUser);
            return cachedUser;
        }

        return null;
    }

    private CachedUser get(String uid) {
        String key = getKey(uid);
        String data = redisTemplate.opsForValue().get(key);
        if (data == null) {
            return null;
        }
        try {
            return objectMapper.readValue(data, CachedUser.class);
        } catch (Exception e) {
            log.error("Failed to deserialize CachedUser", e);
            return null;
        }
    }

    private void set(CachedUser user) {
        String key = getKey(user.uid());
        try {
            String json = objectMapper.writeValueAsString(user);
            redisTemplate.opsForValue().set(key, json, SESSION_TTL, TimeUnit.SECONDS);
        } catch (Exception e) {
            log.error("Failed to cache user session", e);
        }
    }

    private void invalidate(String uid) {
        String key = getKey(uid);
        redisTemplate.delete(key);
    }

    private String getKey(String uid) {
        return SESSION_PREFIX + uid;
    }
}
